sensing/autoware_pointcloud_preprocessor/include/autoware/pointcloud_preprocessor/static_transform_buffer.hpp

```diff
@@ -0,0 +1,137 @@
+
+namespace std
+{
+template <>
+struct hash<std::pair<std::string, std::string>>
+{
+  size_t operator()(const std::pair<std::string, std::string> & p) const
+  {
+    size_t h1 = std::hash<std::string>{}(p.first);
+    size_t h2 = std::hash<std::string>{}(p.second);
+    return h1 ^ (h2 << 1u);
+  }
+};
+}  // namespace std
+
+namespace autoware::pointcloud_preprocessor
+{
+using std::chrono_literals::operator""ms;
+using Key = std::pair<std::string, std::string>;
+struct PairEqual
+{
+  bool operator()(const Key & p1, const Key & p2) const
+  {
+    return p1.first == p2.first && p1.second == p2.second;
+  }
+};
+using TFMap = std::unordered_map<Key, Eigen::Matrix4f, std::hash<Key>, PairEqual>;
+
+class StaticTransformBuffer
+{
+public:
+  StaticTransformBuffer() = default;
+
+  bool get_transform(
+    rclcpp::Node * node, const std::string & target_frame, const std::string & source_frame,
+    Eigen::Matrix4f & eigen_transform)
+  {
+    auto key = std::make_pair(target_frame, source_frame);
+    auto key_inv = std::make_pair(source_frame, target_frame);
+
+    // Check if the transform is already in the buffer
+    auto it = buffer_.find(key);
+    if (it != buffer_.end()) {
+      eigen_transform = it->second;
+      return true;
+    }
+
+    // Check if the inverse transform is already in the buffer
+    auto it_inv = buffer_.find(key_inv);
+    if (it_inv != buffer_.end()) {
+      eigen_transform = it_inv->second.inverse();
+      buffer_[key] = eigen_transform;
+      return true;
+    }
+
+    // Check if transform is needed
+    if (target_frame == source_frame) {
+      eigen_transform = Eigen::Matrix4f::Identity();
+      buffer_[key] = eigen_transform;
+      return true;
+    }
+
+    // Get the transform from the TF tree
+    auto tf_listener = std::make_shared<autoware::universe_utils::TransformListener>(node);
+    auto tf = tf_listener->getTransform(
+      target_frame, source_frame, rclcpp::Time(0), rclcpp::Duration(1000ms));
+    RCLCPP_DEBUG(
+      node->get_logger(), "Trying to enqueue %s -> %s transform to static TFs buffer...",
+      target_frame.c_str(), source_frame.c_str());
+    if (tf == nullptr) {
+      eigen_transform = Eigen::Matrix4f::Identity();
+      return false;
+    }
+    pcl_ros::transformAsMatrix(*tf, eigen_transform);
+    buffer_[key] = eigen_transform;
+    return true;
+  }
+
+  bool transform_pointcloud(
+    rclcpp::Node * node, const std::string & target_frame,
+    const sensor_msgs::msg::PointCloud2 & cloud_in, sensor_msgs::msg::PointCloud2 & cloud_out)
+  {
+    if (
+      pcl::getFieldIndex(cloud_in, "x") == -1 || pcl::getFieldIndex(cloud_in, "y") == -1 ||
+      pcl::getFieldIndex(cloud_in, "z") == -1) {
+      RCLCPP_ERROR(node->get_logger(), "Input pointcloud does not have xyz fields");
+      return false;
+    }
+    Eigen::Matrix4f eigen_transform;
+    if (!get_transform(node, target_frame, cloud_in.header.frame_id, eigen_transform)) {
+      return false;
+    }
```

Comment:
The frame_id coming into the input data may not necessarily differ from the target frame. Therefore, how about including processing that avoids transformation when the input frame and target frame are the same?

Reply:
Please, check `static_tf_buffer_->transform_pointcloud` occurrences within `filter.cpp`. There are conditions which ignore transform call if frames are equal. I didn't change this behavior and I left this as it was. Do you think it will be better to move this logic into StaticTransformBuffer class?

Reply:
@amadeuszsz Thank you for your response.
I understand. So this is intended to be used in filter.cpp.

I am just making a note for the future and not making a request.
I think StaticTransformBuffer would be very useful not only for the pointcloud preprocessor but also for Perception/Localization, so it might be a good idea to move it to a common utility in the future.

Reply:
@yukkysaito 
Sorry, i just read this comment. If we intend to use this implementation out of pointcloud preprocessor, I agree we can skip unnecessary transforms. It will change nothing for pointcloud preprocessor now, but in future we don't need to wrap all transform call around `frame_id` check.
Addressed in https://github.com/autowarefoundation/autoware.universe/pull/8174/commits/d31481c4f32485a6e5a7a62a7eed715fa3df8d1a.